### PR TITLE
don't convert numbers to Strings in GenericData.addPair

### LINKED ISSUE #2419: performance improvements for logging - don't convert numbers to Strings in GenericData.addPair
- modify GenericData to store/retrieve numbers as numbers
- take advantage of this for all places that currently have to convert ints/longs from String to number (eg. in CollectorJsonUtils)

### PR DESCRIPTION
**Pull request by Yushan-Lin**
Fixes #2419

### CODE REVIEW

**Comment by Yushan-Lin** [comment]
#build

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 1]
```diff
@@ -156,42 +165,51 @@ private static String jsonifyFFDC(int maxFieldLength, String wlpUserDir,
...
+                    if (!key.equals(LogFieldConstants.LABEL) && !(key.equals(LogFieldConstants.SOURCEID))
+                        && !(key.equals(LogFieldConstants.DATEOFFIRSTOCCURENCE)) && !(key.equals(LogFieldConstants.COUNT))) {
 
-                if (!key.equals(LogFieldConstants.LABEL) && !(key.equals(LogFieldConstants.SOURCEID))
-                    && !(key.equals(LogFieldConstants.DATEOFFIRSTOCCURENCE)) && !(key.equals(LogFieldConstants.COUNT))) {
+                        if (key.equals(LogFieldConstants.IBM_STACKTRACE)) {
 
-                    if (key.equals(LogFieldConstants.IBM_STACKTRACE)) {
+                            String formattedValue = CollectorJsonHelpers.formatMessage(kvp.getValue(), maxFieldLength);
+                            CollectorJsonHelpers.addToJSON(sb, key, formattedValue, false, true, false, false, kvp.isInteger());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: last parm should be false

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 2]
```diff
@@ -156,42 +165,51 @@ private static String jsonifyFFDC(int maxFieldLength, String wlpUserDir,
...
-                        CollectorJsonHelpers.addToJSON(sb, key, value, false, true, false, false, kvp.isNumber());
+                            String value = null;
+                            if (kvp.isInteger()) {
+                                value = kvp.getIntValue().toString();
+                            } else if (kvp.isLong()) {
+                                value = kvp.getLongValue().toString();
+                            } else {
+                                value = kvp.getValue();
+                            }
+                            CollectorJsonHelpers.addToJSON(sb, key, value, false, true, false, false, kvp.isLong() || kvp.isInteger());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: change last parm to !kvp.isString()

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 3]
```diff
@@ -210,53 +228,62 @@ public static String jsonifyAccess(int maxFieldLength, String wlpUserDir,
...
+                        if (jsonQueryString != null) {
+                            try {
+                                jsonQueryString = URLDecoder.decode(jsonQueryString, "UTF-8");
+                            } catch (UnsupportedEncodingException e) {
+                                // ignore, use the original value;
+                            }
                         }
-                    }
-                    CollectorJsonHelpers.addToJSON(sb, key, jsonQueryString, false, true, false, false, kvp.isNumber());
+                        CollectorJsonHelpers.addToJSON(sb, key, jsonQueryString, false, true, false, false, kvp.isInteger());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: last parm should be false

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 4]
```diff
@@ -210,53 +228,62 @@ public static String jsonifyAccess(int maxFieldLength, String wlpUserDir,
...
 
-                    if (userAgent != null && userAgent.length() > MAX_USER_AGENT_LENGTH) {
-                        userAgent = userAgent.substring(0, MAX_USER_AGENT_LENGTH);
-                    }
+                        if (userAgent != null && userAgent.length() > MAX_USER_AGENT_LENGTH) {
+                            userAgent = userAgent.substring(0, MAX_USER_AGENT_LENGTH);
+                        }
 
-                    CollectorJsonHelpers.addToJSON(sb, key, userAgent, false, false, false, false, kvp.isNumber());
+                        CollectorJsonHelpers.addToJSON(sb, key, userAgent, false, false, false, false, kvp.isInteger());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: last parm should be false

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 5]
```diff
@@ -210,53 +228,62 @@ public static String jsonifyAccess(int maxFieldLength, String wlpUserDir,
...
-                    CollectorJsonHelpers.addToJSON(sb, key, value, false, true, false, false, kvp.isNumber());
+                        String value = null;
+                        if (kvp.isInteger()) {
+                            value = kvp.getIntValue().toString();
+                        } else if (kvp.isLong()) {
+                            value = kvp.getLongValue().toString();
+                        } else {
+                            value = kvp.getValue();
+                        }
+                        CollectorJsonHelpers.addToJSON(sb, key, value, false, true, false, false, kvp.isLong() || kvp.isInteger());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: last parm should be !kvp.isString()

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 6]
```diff
@@ -276,45 +303,54 @@ private static String jsonifyTraceAndMessage(int maxFieldLength, String wlpUserD
...
+                    if (key.equals(LogFieldConstants.SEVERITY)) {
+                    }
 
-                else if (key.equals(LogFieldConstants.MESSAGE)) {
+                    else if (key.equals(LogFieldConstants.MESSAGE)) {
 
-                    String formattedValue = CollectorJsonHelpers.formatMessage(value, maxFieldLength);
-                    CollectorJsonHelpers.addToJSON(sb, key, formattedValue, false, true, false, false, kvp.isNumber());
+                        String formattedValue = CollectorJsonHelpers.formatMessage(kvp.getValue(), maxFieldLength);
+                        CollectorJsonHelpers.addToJSON(sb, key, formattedValue, false, true, false, false, kvp.isInteger());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: last parm should be false

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java [hunk 7]
```diff
@@ -276,45 +303,54 @@ private static String jsonifyTraceAndMessage(int maxFieldLength, String wlpUserD
...
-                    CollectorJsonHelpers.addToJSON(sb, key, value, false, true, false, false, kvp.isNumber());
+                        String value = null;
+                        if (kvp.isInteger()) {
+                            value = kvp.getIntValue().toString();
+                        } else if (kvp.isLong()) {
+                            value = kvp.getLongValue().toString();
+                        } else {
+                            value = kvp.getValue();
+                        }
+                        CollectorJsonHelpers.addToJSON(sb, key, value, false, true, false, false, kvp.isLong() || kvp.isInteger());
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils1_1.java]: last parm should be !kvp.isString()

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java [hunk 8]
```diff
@@ -120,41 +120,52 @@ private static String jsonifyGCEvent(int maxFieldLength, String wlpUserDir,
         ArrayList<Pair> pairs = genData.getPairs();
         KeyValuePair kvp = null;
         String key = null;
-        String value = null;
+//        String value = null;
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java]: see comments in CollectorJsonUtils1_1.java

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePairList.java [hunk 9]
```diff
@@ -24,13 +24,19 @@ public KeyValuePairList() {
...
+    //change to string, long, int
     public void addPair(String key, String value) {
-        KeyValuePair kvp = new KeyValuePair(key, value, KeyValuePair.ValueTypes.STRING);
+        KeyValuePair kvp = new KeyValueStringPair(key, value, KeyValuePair.ValueTypes.STRING);
         keyValuePairs.add(kvp);
     }
 
-    public void addPair(String key, Number value) {
-        KeyValuePair kvp = new KeyValuePair(key, value.toString(), KeyValuePair.ValueTypes.NUMBER);
+    public void addPair(String key, Integer value) {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePairList.java]: should be int value

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePairList.java [hunk 10]
```diff
@@ -24,13 +24,19 @@ public KeyValuePairList() {
...
+        KeyValuePair kvp = new KeyValueIntegerPair(key, value, KeyValuePair.ValueTypes.INTEGER);
+        keyValuePairs.add(kvp);
+    }
+
+    public void addPair(String key, Long value) {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePairList.java]: should be long value

**Comment by Yushan-Lin** [comment]
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_z4G04Cb0EeivM75V9klrxw

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java [hunk 11]
```diff
@@ -120,41 +120,50 @@ private static String jsonifyGCEvent(int maxFieldLength, String wlpUserDir,
         ArrayList<Pair> pairs = genData.getPairs();
         KeyValuePair kvp = null;
         String key = null;
-        String value = null;
 
         StringBuilder sb = CollectorJsonHelpers.startGCJson(hostName, wlpUserDir, serverName);
+        try {
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/CollectorJsonUtils.java]: the new try / catch blocks are no longer needed in this and CollectorJsonUtils1_1.

File: dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePairList.java [hunk 12]
```diff
@@ -24,13 +24,19 @@ public KeyValuePairList() {
         return keyValuePairs;
     }
 
+    //change to string, long, int
```
Comment (donbourne) [inline dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/KeyValuePairList.java]: remove comment

**Comment by Yushan-Lin** [comment]
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_MwZ-YCepEeivM75V9klrxw

**Comment by Yushan-Lin** [comment]
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_qVFkRSepEeivM75V9klrxw

**Pull request merged**